### PR TITLE
Added workflow to publish Docker images to DockerHub automatically

### PR DESCRIPTION
**Pull request by tgaddair**

### CODE REVIEW

**Comment by tgaddair** [comment]
Builds are failing because GitHub runners only have 14GB of disk space. We may need to make use of [this](https://github.com/machulav/ec2-github-runner#customize-hardware-configuration) plugin to use a beefier instance in EC2.

**Comment by tgaddair** [comment]
Only limitation at the moment is that the Docker builds don't run in parallel. This looks to be addressed by https://github.com/machulav/ec2-github-runner/issues/8, but not ETA is given at the moment.

File: build-docker-images.sh [hunk 1]
```diff
@@ -1,41 +0,0 @@
-#!/bin/bash
```
Comment (richardliaw) [inline build-docker-images.sh:1]: hmm any reason to remove this file? this could still help users if they wanted to build themselves right?

Comment (tgaddair) [reply build-docker-images.sh:1]: Let me add a README to help with this.

Comment (tgaddair) [reply build-docker-images.sh:1]: Added README, let me know what you think.

**Review by richardliaw — APPROVED** [review]
Looks fine to me!

File: build-docker-images.sh (hunk 1 above)
Comment (richardliaw) [reply build-docker-images.sh:1]: nice!

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+  stop-runner:
+    name: Stop self-hosted EC2 runner
+    needs:
+      - start-runner # required to get output from the start-runner job
+      - docker # required to wait when the main job is done
+    runs-on: ubuntu-latest
+    if: ${{ always() }} # required to stop the runner even if the error happened in the previous jobs
```
Comment (EnricoMi) [inline .github/workflows/docker.yml:101]: The `${{` and `}}` are not required in `if:` conditions.

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  schedule:
+    - cron: '0 10 * * *' # everyday at 10am
+  push:
+    branches: [ master ]
+    tags: [ 'v*.*.*' ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  start-runner:
```
Comment (EnricoMi) [inline .github/workflows/docker.yml:13]: Why is this running on a self-hosted runner? Because GH runner have too small disks?

File: docker/horovod-ray/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+USER ray
+RUN sudo mkdir -p /horovod && sudo chown ray:users /horovod
+WORKDIR /horovod
+
+# Install Horovod, temporarily using CUDA stubs
+COPY --chown=ray:users . .
+RUN python setup.py sdist && \
+    sudo ldconfig /usr/local/cuda/targets/x86_64-linux/lib/stubs && \
+    HOROVOD_GPU_OPERATIONS=NCCL HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_PYTORCH=1 pip install --no-cache-dir -v $(ls /horovod/dist/horovod-*.tar.gz)[ray] && \
```
Comment (EnricoMi) [inline docker/horovod-ray/Dockerfile:46]: why not `[all-frameworks,ray]`?

File: docs/contributors.rst [hunk 5]
```diff
@@ -339,31 +339,4 @@ Verify that the latest version of Horovod is now available:
 
     $ pip install --upgrade horovod
 
-
```
Comment (EnricoMi) [inline docs/contributors.rst:342]: no replacement here? The content of `docker/README.md`?

File: .github/workflows/docker.yml [hunk 6]
```diff
@@ -0,0 +1,115 @@
+name: docker
+
+on:
+  schedule:
+    - cron: '0 10 * * *' # everyday at 10am
+  push:
+    branches: [ master ]
+    tags: [ 'v*.*.*' ]
+  pull_request:
```
Comment (EnricoMi) [inline .github/workflows/docker.yml:9]: Why on `pull_request` on master? Merging PR into master is covered by `push`.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (EnricoMi) [reply .github/workflows/docker.yml:13]: @tgaddair saw your comment above.

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -0,0 +1,115 @@
+name: docker
+
+on:
+  schedule:
+    - cron: '0 10 * * *' # everyday at 10am
```
Comment (EnricoMi) [inline .github/workflows/docker.yml:5]: since we are publishing with fixed framework versions it would be sufficient to publish when merging into master, not every day, so coupling this to our unit test workflow

File: .github/workflows/docker.yml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      -
+        name: Docker meta
+        id: meta
+        uses: crazy-max/ghaction-docker-meta@v2
+        with:
+          # list of Docker images to use as base name for tags
+          images: |
+            horovod/${{ matrix.docker-image }}
+          # generate Docker tags based on the following events/attributes
+          tags: |
```
Comment (EnricoMi) [inline .github/workflows/docker.yml:64]: Which tags will scheduled master builds get? `nightly` or a date?

File: .github/workflows/docker.yml (hunk 3 above)
Comment (tgaddair) [reply .github/workflows/docker.yml:13]: Yeah, they run out disk pretty quickly, as the image is over 10GB.

File: docker/horovod-ray/Dockerfile (hunk 4 above)
Comment (tgaddair) [reply docker/horovod-ray/Dockerfile:46]: No need for Spark and MXNet for Ray, and PyTorch / TensorFlow should already be installed.

File: docs/contributors.rst (hunk 5 above)
Comment (tgaddair) [reply docs/contributors.rst:342]: Yeah, because this is done automatically through a workflow, we no longer need to perform these steps manually.

File: .github/workflows/docker.yml (hunk 6 above)
Comment (tgaddair) [reply .github/workflows/docker.yml:9]: The `pull_request` allows us to test that build runs successfully (without pushing to DockerHub).

File: docs/contributors.rst (hunk 5 above)
Comment (EnricoMi) [reply docs/contributors.rst:342]: ah, you mean this were the docker release steps, I was always wondering where I could find them...

File: .github/workflows/docker.yml (hunk 7 above)
Comment (tgaddair) [reply .github/workflows/docker.yml:5]: That's mostly true, except for the Ray image, which is updated nightly.  Might be worth finding a way to exclude the others.

File: .github/workflows/docker.yml (hunk 8 above)
Comment (tgaddair) [reply .github/workflows/docker.yml:64]: The cron builds will get the `nightly` tag.

**Comment by EnricoMi** [comment]
The docker meta action adds the following label: `org.opencontainers.image.licenses=NOASSERTION`, happy with that? No idea what that means, though.

File: .github/workflows/docker.yml (hunk 8 above)
Comment (EnricoMi) [reply .github/workflows/docker.yml:64]: good

File: .github/workflows/docker.yml (hunk 2 above)
Comment (tgaddair) [reply .github/workflows/docker.yml:101]: Nice, fixed!

**Comment by tgaddair** [comment]
Hmm, me neither, but seems to be okay.

File: .github/workflows/docker.yml [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        with:
+          mode: start
+          github-token: ${{ secrets.GH_PERSONAL_ACCESS_TOKEN }}
+          ec2-image-id: ami-0759580dedc953d1f
+          ec2-instance-type: c5.9xlarge
+          subnet-id: subnet-0983be43
+          security-group-id: sg-4cba0d08
+
+  docker:
+    if: ${{ github.repository_owner == 'horovod' }}
```
Comment (EnricoMi) [inline .github/workflows/docker.yml:38]: a few more `if:` clauses in the file

**Review by EnricoMi — APPROVED** [review]
Looks good, nice workflow!

File: docs/contributors.rst (hunk 5 above)
Comment (tgaddair) [reply docs/contributors.rst:342]: :)

**Comment by EnricoMi** [comment]
> Only limitation at the moment is that the Docker builds don't run in parallel.

You could spin up one self-hosted instance per matrix value to build in parallel. Easiest way would be to move start-runner and stop-runner steps into docker job. Though it does not improve readability ...

**Comment by tgaddair** [comment]
> > Only limitation at the moment is that the Docker builds don't run in parallel.
> 
> You could spin up one self-hosted instance per matrix value to build in parallel. Easiest way would be to move start-runner and stop-runner steps into docker job. Though it does not improve readability ...

I thought about that, but I don't think it will work, because the two jobs need to execute on different machines. For the `start EC2` job it will run on a normal GitHub machine, while the `docker` job needs to execute on the already initialized EC2 instance.

**Comment by EnricoMi** [comment]
Right, that is tricky. So the start and stop job would need the same matrix, and then there needs to be a way to get the instance name from the respective matrix start sub job.

Is something like this possible?

```
  start-runner:
    id: start-runner-${{ matrix.docker-image }}
    strategy:
      fail-fast: false
      matrix:
        docker-image:
          - horovod
          - horovod-cpu
          - horovod-ray

  docker:
    needs: start-runner # required to start the main job when the runner is ready
    runs-on: ${{ needs.start-runner-${{ matrix.docker-image }}.outputs.label }} # run the job on the newly created runner
    strategy:
      fail-fast: false
      matrix:
        docker-image:
          - horovod
          - horovod-cpu
          - horovod-ray
```

Probably not ...

**Comment by tgaddair** [comment]
I was wondering about that, too. I am not sure, but may be worth discussing in this thread at some point: https://github.com/machulav/ec2-github-runner/issues/8
